core/backend-itwin-client/package.json

```diff
@@ -99,7 +100,9 @@
     "js-base64": "^2.4.5",
     "jsonwebtoken": "^8.5.0",
     "openid-client": "^3.15.3",
-    "semver": "^5.5.0"
+    "semver": "^5.5.0",
+    "agentkeepalive": "^4.1.4",
```

Comment:
are these new dependencies? What do they do?

Reply:
Yes,
1. `agentkeepalive` is use to keep tcp connection alive to server while downloading blocks. Node default http pool closes the connection after it completes the http request. This causes connection to be created and dropped for each block download request. Using above library keep the connection alive and so they are not closed until the file download completes.

2. `proper-lockfile` is use to protect download from two processes or same process attempt to download same file. Its just use ondisk lock file nothing fancy. Previously we have use a method call atomic write. It basically download file with different name (random number) and then change the name to filename when done. To my surprise one of our test open different version of same imodel basically downloaded same set of changeset and checkpoint multiple times. Its goes like this
if (file does not exist)
   start download in temp file 
   atomic rename the file and replace any existing with that name

Each async open call for same imodel will hit above code and would not file changeset file or checkpoint file and start download of exactly same file. At the end which ever complete last just override previous downloaded file.

The same behavior was with multi process using same cache folder downloading the same imodel or version of imodel.

new behavior
1. No matter how many call to download a changeset or checkpoint the new download only download it once but report progress and accept cancel request from any client.
2. Two process downloading same file will simply wait for the main process to finish download and will also report progress but will not attempt to download it again.